1-theory/3-api-contracts.adoc

```diff
@@ -0,0 +1,248 @@
+
+Another popular language is http://json-schema.org/[JSON Schema], which parts of OpenAPI are based on, and mostly compatible with.
+
+.An overly simplified example of a basic API which lists collections and resources of hats.
+[source,yaml]
+----
+openapi: 3.0.2
+info:
+  title: Cat on the Hat API
+  version: 1.0.0
+  description: The API for selling hats with pictures of cats.
+servers:
+  - url: "https://hats.example.com"
+    description: Production server
+  - url: "https://hats-staging.example.com"
+    description: Staging server
+
+paths:
+  /hats:
+    get:
+      description: Returns all hats from the system that the user has access to
+      responses:
+        '200':
+          description: A list of hats.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/hats'
+
+components:
+  schemas:
+    hats:
+      type: array
+      items:
+        $ref: "#/components/schemas/hat"
+
+    hat:
+      type: object
+      properties:
+        id:
+          type: string
+          format: uuid
+        name:
+          type: enum
+          enum:
+            - bowler
+            - top
+            - fedora
+----
+
+OpenAPI is a YAML or JSON based descriptive language which covers endpoints, headers, requests and responses, allows for examples in different mime types, outlines errors, and even lets developers write in potential values, validation rules, etc.
+
+These OpenAPI contract files are usually static, usually written down along with the source code, then sometimes deployed to a file hosting like S3 for folks to use. Some managers want to treat these like business secrets and hide them under lock and key, which makes absolutely no damned sense as they are meta-data only. Most "hackers" could probably figure out that you keep your list of companies under `GET /companies`, so just don't make that a publicly available endpoint and you're gonna be ok. PayPal, Microsoft, and other companies make these OpenAPI contracts available to anyone who wants to download them, and this can help folks integrate with your APIs.
+
+You can imagine an OpenAPI file growing to be rather unwieldy once its got 50+ endpoints and more complex examples, but have no fear you can spread things around in multiple files to make it a lot more DRY (Don't Repeat Yourself) and useful. The first thing to go is usually the `components.schema` definitions, which can be moved to their own files. These files are essentially JSON Schema, so you can have the best of both worlds using both OpenAPI and JSON Schema for various things.
+
+It's quite common to then also make the JSON Schema files available over the wire, by putting links to the static .json files in a HTTP response header. This can provide more information about the object to the API client without needing to figure out how to distribute the files to them ahead of time.
```

Comment:
```suggestion
It's quite common to then also make the JSON Schema files available over the wire, by putting links to the static .json files in an HTTP response header. This can provide more information about the object to the API client without needing to figure out how to distribute the files to them ahead of time.
```